File: targets/widget/LiturgicalCalendar.swift
import Foundation

// Ported 1:1 from src/services/liturgical/LiturgicalThemeService.ts.
// Keep the season enum, label wording, and intensity math identical to the
// TS source so the widget never disagrees with the app about today's season.

enum LiturgicalSeason: String {
    case advent
    case christmas
    case ordinaryTimeEarly = "ordinary-time-early"
    case lent
    case holyWeek = "holy-week"
    case easter
    case pentecost
    case ordinaryTime = "ordinary-time"
}

struct LiturgicalMoment {
    let season: LiturgicalSeason
    /// 0.0 (start of season) -> 1.0 (at the highpoint / end)
    let intensity: Double
    /// Human-readable label e.g. "Week 3 of Advent"
    let label: String
}

private let gregorian: Calendar = {
    var cal = Calendar(identifier: .gregorian)
    cal.timeZone = TimeZone.current
    return cal
}()

// MARK: - Easter calculation (Computus algorithm)

private func easterDate(year: Int) -> Date {
    let a = year % 19
    let b = year / 100
    let c = year % 100
    let d = b / 4
    let e = b % 4
    let f = (b + 8) / 25
    let g = (b - f + 1) / 3
    let h = (19 * a + b - d - g + 15) % 30
    let i = c / 4
    let k = c % 4
    let l = (32 + 2 * e + 2 * i - h - k) % 7
    let m = (a + 11 * h + 22 * l) / 451
    let month = (h + l - 7 * m + 114) / 31 // 1-based
    let day = ((h + l - 7 * m + 114) % 31) + 1
    return dateFrom(year: year, month: month, day: day)
}

private func dateFrom(year: Int, month: Int, day: Int) -> Date {
    var comps = DateComponents()
    comps.year = year
    comps.month = month
    comps.day = day
    return gregorian.date(from: comps)!
}

private func addDays(_ date: Date, _ days: Int) -> Date {
    gregorian.date(byAdding: .day, value: days, to: date)!
}

private func startOfDay(_ date: Date) -> Date {
    gregorian.startOfDay(for: date)
}

private func daysBetween(_ a: Date, _ b: Date) -> Int {
    gregorian.dateComponents([.day], from: startOfDay(a), to: startOfDay(b)).day ?? 0
}

private func clamp(_ value: Double, _ minV: Double, _ maxV: Double) -> Double {
    min(max(value, minV), maxV)
}

private func dayOfWeek(_ date: Date) -> Int {
    // Calendar.component(.weekday) is 1 = Sunday, matching JS's Date#getDay()
    gregorian.component(.weekday, from: date) - 1
}

// MARK: - Advent start: 4th Sunday before Christmas

private func adventStart(year: Int) -> Date {
    let christmas = dateFrom(year: year, month: 12, day: 25)
    let dow = dayOfWeek(christmas) // 0 = Sunday
    let daysToSunday = dow == 0 ? 0 : dow
    let fourthSundayBefore = addDays(christmas, -(daysToSunday + 21))
    return startOfDay(fourthSundayBefore)
}

// MARK: - Main entry point

func getLiturgicalMoment(date: Date = Date()) -> LiturgicalMoment {
    let today = startOfDay(date)
    let year = gregorian.component(.year, from: today)

    let easter = startOfDay(easterDate(year: year))
    let ashWednesday = addDays(easter, -46)
    let palmSunday = addDays(easter, -7)
    let pentecost = addDays(easter, 49)
    let advent = adventStart(year: year)
    let christmas = dateFrom(year: year, month: 12, day: 25)
    let baptismOfTheLord: Date = {
        let epiphany = dateFrom(year: year, month: 1, day: 6)
        let dow = dayOfWeek(epiphany)
        return dow == 0 ? dateFrom(year: year, month: 1, day: 13) : addDays(epiphany, 7 - dow)
    }()

    let prevChristmas = dateFrom(year: year - 1, month: 12, day: 25)

    // Advent (current year)
    if today >= advent && today < christmas {
        let total = daysBetween(advent, christmas)
        let elapsed = daysBetween(advent, today)
        let intensity = clamp(Double(elapsed) / Double(total), 0, 1)
        let weekNum = elapsed / 7 + 1
        return LiturgicalMoment(season: .advent, intensity: intensity, label: "Week \(weekNum) of Advent")
    }

    // Christmas (Dec 25 -> Baptism of the Lord) — sustained celebration, flat intensity
    if today >= christmas || today <= baptismOfTheLord {
        let elapsed = today >= christmas ? daysBetween(christmas, today) : daysBetween(prevChristmas, today)
        let dayNum = elapsed + 1
        return LiturgicalMoment(season: .christmas, intensity: 0.8, label: "Day \(dayNum) of Christmas")
    }

    // Lent
    if today >= ashWednesday && today < palmSunday {
        let total = daysBetween(ashWednesday, palmSunday)
        let elapsed = daysBetween(ashWednesday, today)
        let intensity = clamp(Double(elapsed) / Double(total), 0, 1)
        let weekNum = elapsed / 7 + 1
        return LiturgicalMoment(season: .lent, intensity: intensity, label: "Week \(weekNum) of Lent")
    }

    // Holy Week (Palm Sunday -> Holy Saturday)
    if today >= palmSunday && today < easter {
        let total = daysBetween(palmSunday, easter)
        let elapsed = daysBetween(palmSunday, today)
        let intensity = clamp(0.7 + (Double(elapsed) / Double(total)) * 0.3, 0.7, 1) // already intense, deepens to 1
        let days = ["Palm Sunday", "Monday of Holy Week", "Tuesday of Holy Week",
                    "Wednesday of Holy Week", "Holy Thursday", "Good Friday", "Holy Saturday"]
        let label = elapsed >= 0 && elapsed < days.count ? days[elapsed] : "Holy Week"
        return LiturgicalMoment(season: .holyWeek, intensity: intensity, label: label)
    }

    // Pentecost Sunday
    if daysBetween(today, pentecost) == 0 {
        return LiturgicalMoment(season: .pentecost, intensity: 1.0, label: "Pentecost Sunday")
    }

    // Easter (Easter Sunday -> Pentecost)
    if today >= easter && today < pentecost {
        let total = daysBetween(easter, pentecost)
        let elapsed = daysBetween(easter, today)
        let weekNum = elapsed / 7 + 1
        // Sustained joy (0.8) that builds to full intensity in the final week
        let intensity: Double = elapsed >= total - 7
            ? clamp(0.8 + (Double(elapsed - (total - 7)) / 7) * 0.2, 0.8, 1)
            : 0.8
        return LiturgicalMoment(season: .easter, intensity: intensity, label: "Week \(weekNum) of Easter")
    }

    // Ordinary Time — Early: Baptism of the Lord -> Ash Wednesday
    if today > baptismOfTheLord && today < ashWednesday {
        let elapsed = daysBetween(baptismOfTheLord, today)
        let weekNum = elapsed / 7 + 1
        return LiturgicalMoment(season: .ordinaryTimeEarly, intensity: 0.5, label: "Week \(weekNum) in Ordinary Time")
    }

    // Ordinary Time — Late: Pentecost -> Advent
    let elapsed = daysBetween(pentecost, today)
    let weekNum = elapsed / 7 + 1
    return LiturgicalMoment(season: .ordinaryTime, intensity: 0.5, label: "Week \(weekNum) in Ordinary Time")
}

File: android/app/src/main/java/com/readingdaily/scripture/widget/LiturgicalCalendar.kt
package com.readingdaily.scripture.widget

import java.util.Calendar
import java.util.GregorianCalendar
import kotlin.math.floor
import kotlin.math.max
import kotlin.math.min

// Ported 1:1 from src/services/liturgical/LiturgicalThemeService.ts (see the Swift port at
// targets/widget/LiturgicalCalendar.swift for the iOS equivalent). Keep the season set, label
// wording, and intensity math identical to the TS source so neither widget ever disagrees
// with the app about today's season.

enum class LiturgicalSeason(val id: String) {
    ADVENT("advent"),
    CHRISTMAS("christmas"),
    ORDINARY_TIME_EARLY("ordinary-time-early"),
    LENT("lent"),
    HOLY_WEEK("holy-week"),
    EASTER("easter"),
    PENTECOST("pentecost"),
    ORDINARY_TIME("ordinary-time"),
}

data class LiturgicalMoment(
    val season: LiturgicalSeason,
    /** 0.0 (start of season) -> 1.0 (at the highpoint / end) */
    val intensity: Double,
    /** Human-readable label e.g. "Week 3 of Advent" */
    val label: String,
)

private fun dateAt(year: Int, month1Based: Int, day: Int): Calendar {
    val cal = GregorianCalendar()
    cal.clear()
    cal.set(year, month1Based - 1, day, 0, 0, 0)
    return cal
}

private fun addDays(cal: Calendar, days: Int): Calendar {
    val result = cal.clone() as Calendar
    result.add(Calendar.DAY_OF_MONTH, days)
    return result
}

private fun startOfDay(cal: Calendar): Calendar {
    val result = cal.clone() as Calendar
    result.set(Calendar.HOUR_OF_DAY, 0)
    result.set(Calendar.MINUTE, 0)
    result.set(Calendar.SECOND, 0)
    result.set(Calendar.MILLISECOND, 0)
    return result
}

private fun daysBetween(a: Calendar, b: Calendar): Int {
    val msPerDay = 86400000L
    val diff = startOfDay(b).timeInMillis - startOfDay(a).timeInMillis
    return (diff / msPerDay).toInt()
}

private fun clamp(value: Double, minV: Double, maxV: Double): Double = min(max(value, minV), maxV)

/** Calendar.DAY_OF_WEEK is 1 = Sunday, matching JS's Date#getDay(). */
private fun dayOfWeek(cal: Calendar): Int = cal.get(Calendar.DAY_OF_WEEK) - 1

// Easter calculation (Computus algorithm)
private fun easterDate(year: Int): Calendar {
    val a = year % 19
    val b = year / 100
    val c = year % 100
    val d = b / 4
    val e = b % 4
    val f = (b + 8) / 25
    val g = (b - f + 1) / 3
    val h = (19 * a + b - d - g + 15) % 30
    val i = c / 4
    val k = c % 4
    val l = (32 + 2 * e + 2 * i - h - k) % 7
    val m = (a + 11 * h + 22 * l) / 451
    val month = (h + l - 7 * m + 114) / 31 // 1-based
    val day = ((h + l - 7 * m + 114) % 31) + 1
    return dateAt(year, month, day)
}

// Advent start: 4th Sunday before Christmas
private fun adventStart(year: Int): Calendar {
    val christmas = dateAt(year, 12, 25)
    val dow = dayOfWeek(christmas) // 0 = Sunday
    val daysToSunday = if (dow == 0) 0 else dow
    return startOfDay(addDays(christmas, -(daysToSunday + 21)))
}

fun getLiturgicalMoment(date: Calendar = GregorianCalendar()): LiturgicalMoment {
    val today = startOfDay(date)
    val year = today.get(Calendar.YEAR)

    val easter = startOfDay(easterDate(year))
    val ashWednesday = addDays(easter, -46)
    val palmSunday = addDays(easter, -7)
    val pentecost = addDays(easter, 49)
    val advent = adventStart(year)
    val christmas = dateAt(year, 12, 25)
    val baptismOfTheLord: Calendar = run {
        val epiphany = dateAt(year, 1, 6)
        val dow = dayOfWeek(epiphany)
        if (dow == 0) dateAt(year, 1, 13) else addDays(epiphany, 7 - dow)
    }
    val prevChristmas = dateAt(year - 1, 12, 25)

    // Advent (current year)
    if (today >= advent && today < christmas) {
        val total = daysBetween(advent, christmas)
        val elapsed = daysBetween(advent, today)
        val intensity = clamp(elapsed.toDouble() / total, 0.0, 1.0)
        val weekNum = elapsed / 7 + 1
        return LiturgicalMoment(LiturgicalSeason.ADVENT, intensity, "Week $weekNum of Advent")
    }

    // Christmas (Dec 25 -> Baptism of the Lord) — sustained celebration, flat intensity
    if (today >= christmas || today <= baptismOfTheLord) {
        val elapsed = if (today >= christmas) daysBetween(christmas, today) else daysBetween(prevChristmas, today)
        val dayNum = elapsed + 1
        return LiturgicalMoment(LiturgicalSeason.CHRISTMAS, 0.8, "Day $dayNum of Christmas")
    }

    // Lent
    if (today >= ashWednesday && today < palmSunday) {
        val total = daysBetween(ashWednesday, palmSunday)
        val elapsed = daysBetween(ashWednesday, today)
        val intensity = clamp(elapsed.toDouble() / total, 0.0, 1.0)
        val weekNum = elapsed / 7 + 1
        return LiturgicalMoment(LiturgicalSeason.LENT, intensity, "Week $weekNum of Lent")
    }

    // Holy Week (Palm Sunday -> Holy Saturday)
    if (today >= palmSunday && today < easter) {
        val total = daysBetween(palmSunday, easter)
        val elapsed = daysBetween(palmSunday, today)
        val intensity = clamp(0.7 + (elapsed.toDouble() / total) * 0.3, 0.7, 1.0)
        val days = listOf(
            "Palm Sunday", "Monday of Holy Week", "Tuesday of Holy Week",
            "Wednesday of Holy Week", "Holy Thursday", "Good Friday", "Holy Saturday"
        )
        val label = if (elapsed in days.indices) days[elapsed] else "Holy Week"
        return LiturgicalMoment(LiturgicalSeason.HOLY_WEEK, intensity, label)
    }

    // Pentecost Sunday
    if (daysBetween(today, pentecost) == 0) {
        return LiturgicalMoment(LiturgicalSeason.PENTECOST, 1.0, "Pentecost Sunday")
    }

    // Easter (Easter Sunday -> Pentecost)
    if (today >= easter && today < pentecost) {
        val total = daysBetween(easter, pentecost)
        val elapsed = daysBetween(easter, today)
        val weekNum = elapsed / 7 + 1
        val intensity = if (elapsed >= total - 7) {
            clamp(0.8 + ((elapsed - (total - 7)).toDouble() / 7) * 0.2, 0.8, 1.0)
        } else {
            0.8
        }
        return LiturgicalMoment(LiturgicalSeason.EASTER, intensity, "Week $weekNum of Easter")
    }

    // Ordinary Time — Early: Baptism of the Lord -> Ash Wednesday
    if (today > baptismOfTheLord && today < ashWednesday) {
        val elapsed = daysBetween(baptismOfTheLord, today)
        val weekNum = elapsed / 7 + 1
        return LiturgicalMoment(LiturgicalSeason.ORDINARY_TIME_EARLY, 0.5, "Week $weekNum in Ordinary Time")
    }

    // Ordinary Time — Late: Pentecost -> Advent
    val elapsed = daysBetween(pentecost, today)
    val weekNum = elapsed / 7 + 1
    return LiturgicalMoment(LiturgicalSeason.ORDINARY_TIME, 0.5, "Week $weekNum in Ordinary Time")
}

private operator fun Calendar.compareTo(other: Calendar): Int = this.timeInMillis.compareTo(other.timeInMillis)
